File: rust/crates/fantastic-kernel/src/kernel.rs
//! Shared kernel context — flat agent index, inbox queues, bundle
//! registry, root pointer.
//!
//! The [`Kernel`] is `Arc`-cloneable and `Send + Sync` so handler
//! closures can capture it without contention. Concurrent paths use
//! lock-free `DashMap`s for the agent + inbox lookups; the root
//! pointer is an `ArcSwap` so swaps during boot are observable
//! atomically by every reader.

use crate::agent::{Agent, AgentId};
use crate::bundle::BundleRegistry;
use crate::errors::{KernelError, KernelResult};
use crate::state::{KernelState, CURRENT_VERSION};
use crate::storage::StorageMode;
use arc_swap::ArcSwapOption;
use dashmap::DashMap;
use serde_json::Value;
use std::path::PathBuf;
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::{Arc, RwLock};
use tokio::sync::mpsc;

/// Bounded inbox capacity per agent — matches Python's `INBOX_BOUND`.
/// 256 fits a chat-paced workload comfortably and bounds memory on a
/// runaway emitter. Tunable via [`Kernel::new_with_inbox_bound`].
pub const DEFAULT_INBOX_BOUND: usize = 256;

/// A subscriber callback fired on every state event (`send` / `emit`
/// / `removed` / etc.). Synchronous, called inline — keep it cheap;
/// off-thread the work if it might block.
pub type StateSubscriber = Arc<dyn Fn(&Value) + Send + Sync>;

/// Opaque token returned by [`Kernel::add_state_subscriber`]; pass to
/// [`Kernel::remove_state_subscriber`] to detach.
///
/// The inner `u64` is `pub` so embedding consumers can round-trip the
/// value through a primitive type (e.g. a JSON string) across a
/// process boundary. Treat it as opaque from Rust callers — its
/// numeric value carries no meaning beyond "token issued by this
/// kernel".
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct SubscriberToken(pub u64);

/// Shared kernel context.
pub struct Kernel {
    /// Flat global id → Agent. The only routing table.
    pub agents: DashMap<AgentId, Arc<Agent>>,
    /// Per-id inbox (agents + synthetic browser client ids).
    pub inboxes: DashMap<AgentId, mpsc::Sender<Value>>,
    /// Active state-event taps (telemetry, debugging). Keyed by token
    /// for clean detach.
    state_subscribers: RwLock<Vec<(SubscriberToken, StateSubscriber)>>,
    /// Monotonic counter for [`SubscriberToken`].
    next_subscriber: AtomicU64,
    /// Compile-time bundle registry — caller fills before bootstrap.
    pub bundles: BundleRegistry,
    /// The root agent (`id="core"` by convention). Set during bootstrap.
    pub root: ArcSwapOption<Agent>,
    /// Per-inbox bound; same value applied to every channel.
    pub inbox_bound: usize,
    /// State medium. [`StorageMode::Disk`] auto-flushes to
    /// `<workdir>/.fantastic/state.json` on every mutation;
    /// [`StorageMode::InMemory`] never touches the filesystem. See
    /// [`Self::save`] / [`Self::load`] for the on-demand API both
    /// modes share.
    pub storage: StorageMode,
}

impl Default for Kernel {
    fn default() -> Self {
        Self::new()
    }
}

impl Kernel {
    /// Empty kernel — no agents, no root, default inbox bound,
    /// [`StorageMode::InMemory`]. The `Disk` mode is opt-in via
    /// [`Self::with_storage`]; the no-arg ctor keeps the existing
    /// per-test default (these tests never persist).
    pub fn new() -> Self {
        Self::with_storage(StorageMode::InMemory)
    }

    /// Empty kernel with a specific [`StorageMode`]. Used by
    /// [`crate::bootstrap::bootstrap`] to pick `Disk` or `InMemory`
    /// based on caller options.
    pub fn with_storage(storage: StorageMode) -> Self {
        Self::with_storage_and_inbox(storage, DEFAULT_INBOX_BOUND)
    }

    /// Empty kernel with a custom inbox bound. Use for testing
    /// back-pressure paths.
    pub fn new_with_inbox_bound(bound: usize) -> Self {
        Self::with_storage_and_inbox(StorageMode::InMemory, bound)
    }

    /// Empty kernel with both a [`StorageMode`] and a custom inbox
    /// bound — the underlying ctor every other variant funnels into.
    pub fn with_storage_and_inbox(storage: StorageMode, bound: usize) -> Self {
        Self {
            agents: DashMap::new(),
            inboxes: DashMap::new(),
            state_subscribers: RwLock::new(Vec::new()),
            next_subscriber: AtomicU64::new(1),
            bundles: BundleRegistry::new(),
            root: ArcSwapOption::const_empty(),
            inbox_bound: bound,
            storage,
        }
    }

    /// Register an agent in the flat index + auto-vivify its inbox.
    /// Returns the bound channel receiver so the caller can drain it
    /// (typically a fanout loop or a per-bundle handler task).
    pub fn register(&self, agent: Arc<Agent>) -> mpsc::Receiver<Value> {
        let (tx, rx) = mpsc::channel(self.inbox_bound);
        self.inboxes.insert(agent.id.clone(), tx);
        self.agents.insert(agent.id.clone(), agent);
        rx
    }

    /// Unregister an agent + drop its inbox. Caller has already run
    /// the bundle's `on_delete` hook + rmtreed the dir.
    pub fn unregister(&self, id: &AgentId) {
        self.inboxes.remove(id);
        self.agents.remove(id);
    }

    /// Cheap clone of the current root pointer (`None` until
    /// bootstrap completes).
    pub fn root(&self) -> Option<Arc<Agent>> {
        self.root.load_full()
    }

    /// Swap the root pointer atomically. Bootstrap calls this once.
    pub fn set_root(&self, root: Arc<Agent>) {
        self.root.store(Some(root));
    }

    /// Push a state event to every subscriber. Keep callbacks short —
    /// they run on the dispatching task.
    pub fn publish_state(&self, event: &Value) {
        let subs = self.state_subscribers.read().expect("state subs poisoned");
        for (_token, cb) in subs.iter() {
            cb(event);
        }
    }

    /// Register a state-event tap. Returns a token the caller can
    /// pass to [`Self::remove_state_subscriber`] to detach.
    pub fn add_state_subscriber(&self, cb: StateSubscriber) -> SubscriberToken {
        let token = SubscriberToken(self.next_subscriber.fetch_add(1, Ordering::SeqCst));
        self.state_subscribers
            .write()
            .expect("state subs poisoned")
            .push((token, cb));
        token
    }

    /// Detach a previously-registered subscriber. No-op if `token`
    /// isn't (or no longer is) registered.
    pub fn remove_state_subscriber(&self, token: SubscriberToken) {
        self.state_subscribers
            .write()
            .expect("state subs poisoned")
            .retain(|(t, _)| *t != token);
    }

    // ── save / load — the foundation primitive both storage modes share ──

    /// Snapshot the live kernel state as a [`KernelState`] value.
    /// Both [`StorageMode::Disk`] and [`StorageMode::InMemory`]
    /// kernels produce equal output for equal in-memory state — the
    /// medium difference is irrelevant here.
    ///
    /// Pure read; iterates `self.agents`, builds the snapshot in
    /// id-sorted order so [`Self::save_json`] is byte-deterministic.
    /// Ephemeral agents are skipped (they're per-process composition,
    /// never round-trip through a snapshot).
    pub fn save(&self) -> KernelState {
        let mut agents: Vec<_> = self
            .agents
            .iter()
            .filter(|e| !e.value().ephemeral)
            .map(|e| e.value().record())
            .collect();
        agents.sort_by(|a, b| a.id.cmp(&b.id));
        KernelState {
            version: CURRENT_VERSION,
            agents,
        }
    }

    /// [`Self::save`] serialized to JSON. Used by [`Self::maybe_flush`]
    /// (Disk auto-sync) and by the embedding API.
    /// Output is byte-deterministic for equal kernel states.
    pub fn save_json(&self) -> String {
        serde_json::to_string(&self.save()).expect("KernelState is JSON-serializable")
    }

    /// Replace this kernel's agent tree with `state`. Drops every
    /// currently-registered agent (including their inboxes), then
    /// rebuilds from the snapshot.
    ///
    /// Does NOT fire `boot` hooks — matches the existing
    /// `persistence::load_children` semantics. The bootstrap caller
    /// dispatches `{type:"boot"}` to whichever agents need it
    /// (typically the web agent, or a bundle that boots paired agents,
    /// etc.). For brain-kernel-style snapshot reload, the consumer
    /// can iterate `kernel.agents` post-load and fire boot themselves
    /// if needed.
    ///
    /// Weak-load: agents whose `handler_module` isn't in this
    /// kernel's [`BundleRegistry`] are logged + skipped along with
    /// their entire subtree (a parent's missing bundle implies all
    /// its descendants are unreachable). The record is dropped from
    /// the loaded set; subsequent [`Self::save`] won't see it.
    ///
    /// Errors:
    /// - [`KernelError::InvalidSnapshot`] if `state.version` exceeds
    ///   [`CURRENT_VERSION`], if the snapshot has no root, or if any
    ///   record's `parent_id` doesn't point at a record in the
    ///   snapshot.
    pub fn load(&self, state: KernelState) -> KernelResult<()> {
        if state.version > CURRENT_VERSION {
            return Err(KernelError::InvalidSnapshot(format!(
                "snapshot version {} exceeds this kernel's max ({})",
                state.version, CURRENT_VERSION
            )));
        }

        // Validate: every parent_id must point at a record in the
        // snapshot, exactly one record must have parent_id == None
        // (the root), every id must be unique.
        let mut roots = 0;
        let mut ids_seen: std::collections::HashSet<&str> = std::collections::HashSet::new();
        for rec in &state.agents {
            if !ids_seen.insert(rec.id.as_str()) {
                return Err(KernelError::InvalidSnapshot(format!(
                    "duplicate agent id {:?} in snapshot",
                    rec.id
                )));
            }
            if rec.parent_id.is_none() {
                roots += 1;
            }
        }
        if roots == 0 {
            return Err(KernelError::InvalidSnapshot(
                "snapshot has no root (no record with parent_id == null)".to_string(),
            ));
        }
        if roots > 1 {
            return Err(KernelError::InvalidSnapshot(format!(
                "snapshot has {roots} roots; expected exactly one"
            )));
        }
        for rec in &state.agents {
            if let Some(pid) = &rec.parent_id {
                if !ids_seen.contains(pid.as_str()) {
                    return Err(KernelError::InvalidSnapshot(format!(
                        "agent {:?} parent {:?} not in snapshot",
                        rec.id, pid
                    )));
                }
            }
        }

        // Drop the current tree. Inboxes close as their senders are
        // dropped from the DashMap (subscribers see the channel
        // closing and clean up).
        self.agents.clear();
        self.inboxes.clear();
        self.root.store(None);

        // Compute the per-agent on-disk root_path. Disk mode roots at
        // `<workdir>/.fantastic/`; children nest under `agents/<id>/`.
        // InMemory mode uses an empty PathBuf (never read).
        let disk_root: Option<PathBuf> = self.storage.workdir().map(|w| w.join(".fantastic"));

        // Two-pass: build all Agent structs first (parent_id strings
        // only), then wire parent/child Arc references. The
        // weak-load skip happens in the first pass — a record whose
        // handler_module isn't registered is dropped, along with any
        // descendant whose ancestor chain hits the dropped record.
        let mut by_id: std::collections::HashMap<String, Arc<Agent>> =
            std::collections::HashMap::new();
        let mut skip_ids: std::collections::HashSet<String> = std::collections::HashSet::new();

        // Topologically order: roots first, then BFS by ancestor depth.
        // Simpler approach: iterate; if a record's parent_id is a
        // skipped id, skip this one too. Repeat until no progress.
        // For sane snapshots (≤ a few thousand records) the cost is
        // O(N * depth), negligible.
        let mut pending: Vec<&crate::agent::AgentRecord> = state.agents.iter().collect();
        // Sort so parents always precede children when parent_id is
        // None or refers to an earlier record. Simple stable sort by
        // (parent_depth, id) — implemented by a multi-pass walk.
        let mut made_progress = true;
        while made_progress && !pending.is_empty() {
            made_progress = false;
            let mut next: Vec<&crate::agent::AgentRecord> = Vec::new();
            for rec in pending.drain(..) {
                let parent_id = rec.parent_id.clone();
                if let Some(pid) = parent_id.as_ref() {
                    if skip_ids.contains(pid) {
                        // Cascade skip — parent was dropped (weak-load).
                        skip_ids.insert(rec.id.clone());
                        tracing::warn!(
                            agent = %rec.id,
                            ancestor = %pid,
                            "skipping agent during load (ancestor dropped)",
                        );
                        made_progress = true;
                        continue;
                    }
                    if !by_id.contains_key(pid) {
                        // Parent not yet processed — re-queue.
                        next.push(rec);
                        continue;
                    }
                }

                // Weak-load: refuse this record if its handler_module
                // isn't registered. Root has no handler_module so
                // always loads.
                if let Some(hm) = rec.handler_module.as_deref() {
                    if self.bundles.get(hm).is_none() {
                        eprintln!(
                            "[kernel] skipping agent {}: bundle {} not installed in this runtime",
                            rec.id, hm,
                        );
                        skip_ids.insert(rec.id.clone());
                        made_progress = true;
                        continue;
                    }
                }

                // Construct the root_path. Disk mode: agent path is
                // `<workdir>/.fantastic[/agents/<id>]*` walking up the
                // parent chain. InMemory mode: empty PathBuf.
                let root_path = match (&disk_root, rec.parent_id.as_ref()) {
                    // Disk root.
                    (Some(disk), None) => disk.clone(),
                    // Disk child — reuse parent's already-composed
                    // root_path so we get `<disk>/agents/<id1>/agents/<id2>/...`.
                    (Some(_), Some(pid)) => by_id
                        .get(pid)
                        .expect("parent processed before child")
                        .children_dir()
                        .join(&rec.id),
                    // InMemory — root_path is never read; empty sentinel.
                    (None, _) => PathBuf::new(),
                };

                let agent = Agent::new(
                    AgentId::from(rec.id.as_str()),
                    rec.handler_module.clone(),
                    rec.parent_id.as_ref().map(|s| AgentId::from(s.as_str())),
                    rec.meta.clone(),
                    root_path,
                    false, // not ephemeral — ephemeral agents are never in a snapshot
                );
                let _rx = self.register(Arc::clone(&agent));
                by_id.insert(rec.id.clone(), Arc::clone(&agent));

                // Wire parent → child Arc.
                if let Some(pid) = &rec.parent_id {
                    if let Some(parent_arc) = by_id.get(pid) {
                        parent_arc
                            .children
                            .insert(agent.id.clone(), Arc::clone(&agent));
                    }
                } else {
                    self.set_root(Arc::clone(&agent));
                }

                made_progress = true;
            }
            pending = next;
        }

        if !pending.is_empty() {
            // Cycle or orphan whose parent never appeared (shouldn't
            // happen — validation above caught this).
            return Err(KernelError::InvalidSnapshot(format!(
                "could not resolve {} agents during load (cycle?)",
                pending.len()
            )));
        }

        Ok(())
    }

    /// JSON-string form of [`Self::load`]. Parses then loads. Used by
    /// the embedding API and the Disk-mode boot path.
    pub fn load_json(&self, json: &str) -> KernelResult<()> {
        let state: KernelState = serde_json::from_str(json)
            .map_err(|e| KernelError::InvalidSnapshot(format!("parse: {e}")))?;
        self.load(state)
    }

    /// Synchronous read of every loaded agent's identity + display
    /// name. Mirrors Python's `Kernel.state_snapshot` — used by new
    /// `state_subscribe` subscribers to bootstrap their agent view
    /// before the first event arrives. No queue puts, no fanout.
    ///
    /// Each entry: `{agent_id, name, backlog: 0}`. `backlog` is the
    /// number of in-flight handler dispatches; Rust doesn't track
    /// this counter today, so we report 0 (consumers that rely on
    /// it for ordering can recompute from observed traffic). Matches
    /// Python's wire shape so existing telemetry pane consumers work.
    pub fn state_snapshot(&self) -> Vec<Value> {
        let mut out: Vec<Value> = self
            .agents
            .iter()
            .map(|entry| {
                let a = entry.value();
                let name = a.display_name().unwrap_or_else(|| a.id.0.clone());
                serde_json::json!({
                    "agent_id": a.id.0,
                    "name": name,
                    "backlog": 0,
                })
            })
            .collect();
        out.sort_by(|a, b| {
            a.get("agent_id")
                .and_then(Value::as_str)
                .unwrap_or("")
                .cmp(b.get("agent_id").and_then(Value::as_str).unwrap_or(""))
        });
        out
    }
}

#[cfg(test)]
mod tests;

File: swift/Sources/FantasticKernel/Kernel.swift
// Kernel — the substrate's central actor.
//
// Mirrors Rust's `fantastic_kernel::Kernel`. Owns the flat agents
// map (routing table), per-agent inbox channels, state event
// subscribers, the bundle registry, and the storage mode.
//
// `send(target, payload)` → resolves target → dispatches via system
// verb table OR the agent's bundle → publishes state event +
// watcher fanout → returns reply.
//
// `emit(target, payload)` → appends to target's inbox + publishes
// state event + fans out to watchers. No dispatch.

import FantasticJSON
import Foundation
import OrderedCollections

// ── Current sender task-local ─────────────────────────────────────

/// Task-local that names the agent currently dispatching. Read by
/// `Kernel.send` to attribute state events; written by `send_json_as`
/// + by `Kernel.send` itself before invoking the bundle handler
/// (so nested sends attribute to their target). Mirrors Rust's
/// `CURRENT_SENDER` task-local.
public enum KernelTaskLocals {
    @TaskLocal public static var currentSender: AgentId? = nil
}

/// Subscriber token returned by `Kernel.subscribe`; pass to
/// `Kernel.unsubscribe` to detach.
public struct SubscriberToken: Sendable, Hashable {
    public let value: UInt64
    init(_ value: UInt64) { self.value = value }
}

public typealias StateSubscriber = @Sendable (JSON) -> Void

public final class Kernel: @unchecked Sendable {
    /// The flat agents map. Lock-protected; readers snapshot via
    /// `agent(id:)`. Mirrors Rust's `DashMap<AgentId, Arc<Agent>>`.
    private let agentsLock = NSLock()
    private var _agents: [AgentId: Agent] = [:]

    /// Per-agent inbox stream-continuations. `emit` appends to
    /// these; watchers fan out by reading their own inbox stream.
    private let inboxLock = NSLock()
    private var _inboxes: [AgentId: AsyncStream<JSON>.Continuation] = [:]
    private var _inboxStreams: [AgentId: AsyncStream<JSON>] = [:]

    /// State subscribers — opaque tokens + callback closures.
    private let subscriberLock = NSLock()
    private var _subscribers: [(SubscriberToken, StateSubscriber)] = []
    private var nextSubscriberId: UInt64 = 1

    /// Atomic root agent reference. Set by `setRoot`; read by `root`.
    private let rootLock = NSLock()
    private var _root: Agent?

    /// HTTP port the listener is bound to (0 if no listener). Set
    /// by phase 8B when the Hummingbird listener binds. Exposed via
    /// `httpPort()` shim in `PublicAPI.swift`.
    let httpPortLock = NSLock()
    var _httpPort: UInt16 = 0

    public let bundles: BundleRegistry
    public let storage: StorageMode
    public let inboxBound: Int

    public init(
        storage: StorageMode = .inMemory,
        bundles: BundleRegistry = BundleRegistry(),
        inboxBound: Int = 256
    ) {
        self.storage = storage
        self.bundles = bundles
        self.inboxBound = inboxBound
    }

    // ── Agent map ───────────────────────────────────────────────

    /// Register `agent` in the flat routing map and create its inbox
    /// stream. Returns the stream the caller can iterate to read
    /// inbox messages. Mirrors `Rust Kernel::register` (which returns
    /// `mpsc::Receiver<Value>`).
    @discardableResult
    public func register(_ agent: Agent) -> AsyncStream<JSON> {
        agentsLock.lock()
        _agents[agent.id] = agent
        agentsLock.unlock()
        return ensureInbox(agent.id)
    }

    /// Snapshot the registered agent for `id`, if any.
    public func agent(_ id: AgentId) -> Agent? {
        agentsLock.lock()
        defer { agentsLock.unlock() }
        return _agents[id]
    }

    /// All registered agents in insertion order.
    public func allAgents() -> [Agent] {
        agentsLock.lock()
        defer { agentsLock.unlock() }
        return Array(_agents.values)
    }

    /// Remove an agent from the routing map + drop its inbox.
    /// Used by cascade delete.
    func unregister(_ id: AgentId) {
        agentsLock.lock()
        _agents[id] = nil
        agentsLock.unlock()
        inboxLock.lock()
        _inboxes[id]?.finish()
        _inboxes[id] = nil
        _inboxStreams[id] = nil
        inboxLock.unlock()
    }

    public func setRoot(_ agent: Agent) {
        rootLock.lock()
        _root = agent
        rootLock.unlock()
    }

    public var root: Agent? {
        rootLock.lock()
        defer { rootLock.unlock() }
        return _root
    }

    // ── Inbox + watchers ────────────────────────────────────────

    @discardableResult
    public func ensureInbox(_ id: AgentId) -> AsyncStream<JSON> {
        inboxLock.lock()
        defer { inboxLock.unlock() }
        if let stream = _inboxStreams[id] {
            return stream
        }
        var continuation: AsyncStream<JSON>.Continuation!
        let stream = AsyncStream<JSON>(
            bufferingPolicy: .bufferingNewest(inboxBound)
        ) { c in
            continuation = c
        }
        _inboxes[id] = continuation
        _inboxStreams[id] = stream
        return stream
    }

    func deliverToInbox(_ id: AgentId, _ payload: JSON) {
        inboxLock.lock()
        let continuation = _inboxes[id]
        inboxLock.unlock()
        if let continuation = continuation {
            _ = continuation.yield(payload)
        } else {
            // Auto-vivify for synthetic ids (browser clients), then deliver.
            _ = ensureInbox(id)
            inboxLock.lock()
            let c = _inboxes[id]
            inboxLock.unlock()
            _ = c?.yield(payload)
        }
    }

    /// Register `watcherId` as an observer of `srcId`'s inbox.
    public func watch(src: AgentId, watcher: AgentId) {
        if let src = agent(src) {
            src.addWatcher(watcher)
        }
        ensureInbox(watcher)
    }

    public func unwatch(src: AgentId, watcher: AgentId) {
        if let src = agent(src) {
            src.removeWatcher(watcher)
        }
    }

    func fanoutToWatchers(_ target: Agent, _ payload: JSON) {
        for w in target.watcherIds() {
            deliverToInbox(w, payload)
        }
    }

    // ── State events ────────────────────────────────────────────

    @discardableResult
    public func subscribe(_ cb: @escaping StateSubscriber) -> SubscriberToken {
        subscriberLock.lock()
        let id = nextSubscriberId
        nextSubscriberId += 1
        let token = SubscriberToken(id)
        _subscribers.append((token, cb))
        subscriberLock.unlock()
        return token
    }

    public func unsubscribe(_ token: SubscriberToken) {
        subscriberLock.lock()
        defer { subscriberLock.unlock() }
        _subscribers.removeAll { $0.0 == token }
    }

    func publishState(_ event: JSON) {
        subscriberLock.lock()
        let subs = _subscribers
        subscriberLock.unlock()
        for (_, cb) in subs {
            cb(event)
        }
    }
}
